airflow/operators/python_operator.py

```diff
@@ -145,3 +158,196 @@ def execute(self, context):
+                else:
+                    pickle.dump(arg_dict, f)
+
+    def _read_result(self, output_filename):
+        if os.stat(output_filename).st_size == 0:
+            return None
+        with open(output_filename) as f:
+            try:
+                if self.use_dill:
+                    return dill.load(f)
+                else:
+                    return pickle.load(f)
+            except ValueError:
+                logging.error("Error deserializing result. Note that result deserialization " +
+                              "is not supported across major Python versions.")
+                raise
+
+    def _write_script(self, script_filename):
+        with open(script_filename, 'w') as f:
+            python_code = self._generate_python_code()
+            logging.debug('Writing code to file\n{}'.format(python_code))
+            f.write(python_code)
+
+    def _generate_virtualenv_cmd(self, tmp_dir):
+        cmd = ['virtualenv', tmp_dir]
+        if self.system_site_packages:
+            cmd.append('--system-site-packages')
+        if self.python_version is not None:
+            cmd.append('--python=python{}'.format(self.python_version))
+        return cmd
+
+    def _generate_pip_install_cmd(self, tmp_dir):
+        if len(self.requirements) == 0:
+            return []
+        else:
+            # direct path alleviates need to activate
+            cmd = ['{}/bin/pip'.format(tmp_dir), 'install']
+            return cmd + self.requirements
+
+    def _generate_python_cmd(self, tmp_dir, script_filename, input_filename, output_filename, string_args_filename):
+        # direct path alleviates need to activate
+        return ['{}/bin/python'.format(tmp_dir), script_filename, input_filename, output_filename, string_args_filename]
+            
+    def _generate_python_code(self):
+        if self.use_dill:
+            pickling_library = 'dill'
+        else:
+            pickling_library = 'pickle'
+        fn = self.python_callable
+        # dont try to read pickle if we didnt pass anything
+        if self._pass_op_args():
+            load_args_line = 'with open(sys.argv[1]) as f: arg_dict = {}.load(f)'.format(pickling_library)
+        else:
+            load_args_line = 'arg_dict = {"args": [], "kwargs": {}}'
+
+        # no indents in original code so we can accept any type of indents in the original function
+        # we deserialize args, call function, serialize result if necessary
+        return dedent("""\
+        import {pickling_library}
+        import sys
+        {}
+        args = arg_dict["args"]
+        kwargs = arg_dict["kwargs"]
+        with open(sys.argv[3]) as f: virtualenv_string_args = list(map(lambda x: x.strip(), list(f)))
+        {}
+        res = {}(*args, **kwargs)
+        with open(sys.argv[2], 'wb') as f: res is not None and {pickling_library}.dump(res, f)
+        """).format(load_args_line, dedent(inspect.getsource(fn)), fn.__name__, pickling_library=pickling_library)
```

Comment:
would be nice to use kwargs across here and reference the keys everywhere `{}` is used